include/app_storage.h

```diff
@@ -0,0 +1,81 @@
+/*****************************************************************************
+ *   (c) 2024 Ledger SAS.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *****************************************************************************/
+
+#pragma once
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#ifndef HAVE_BOLOS
+/* "app_storage_data.h" needs to be created in all apps including this file */
+#include "app_storage_data.h"
+#endif  // HAVE_BOLOS
+
+///< bit to indicate in properties that the application storage contains settings
+#define APP_STORAGE_PROP_SETTINGS (1 << 0)
+///< bit to indicate in properties that the application contains data
+#define APP_STORAGE_PROP_DATA     (1 << 1)
+
+/**
+ * @brief Structure defining the header of NVRAM
+ *
+ */
+typedef struct app_storage_header_s {
+    char     tag[4];          ///< ['S','T','O','R'] array, when properly initialized
+    uint32_t size;            ///< size in bytes of the data (app_storage_data_t structure)
+    uint16_t struct_version;  ///< version of the structure of data (to be set once at first
+                              ///< application start-up)
+    uint16_t properties;      ///< used as a bitfield to set properties, like: contains settings,
+                              ///< contains sensitive data
+    uint32_t data_version;    ///< version of the content of data (to be updated every time data are
+                              ///< updated)
+} app_storage_header_t;
+
+/**
+ * @brief Structure defining the NVRAM
+ *
+ */
+typedef struct app_storage_s {
+    app_storage_header_t header;  ///< header describing the data
+#ifndef HAVE_BOLOS
+    app_storage_data_t
+        data;  ///< application data, app_storage_data_t must be defined in "nvram_data.h" file
```

Comment:
Why defining it in nvram_data.h? It could be a simple pointer, no?
What happens if the app doesn't define any structure then?

Reply:
Fixed, it is `app_storage_data.h`.
The struct `app_storage_t` is actually written in NVRAM, thus we need to have it contiguous in memory (header+data).
If an application enable in its Makefile the app storage (`ENABLE_APP_STORAGE`), then it must provide a header file named `app_storage_data.h` that defines this `app_storage_data_t` struct otherwise it will not compile.

Reply:
Thanks.
Is this already defined somewhere in a doc? At least on Confluence, maybe also in Dev Portal?